_gsocblogs/2022/blog_ROOT_JunZhang.md

```diff
@@ -12,24 +13,80 @@ intro: |
+* [Current Working Status of Optimize Modules Usage For Root](https://compiler-research.org/assets/presentations/CaaS_Weekly_07_09_2022_Jun_Zhang_Current_Working_Status_of_Optimize_Modules_Usage_For_ROOT.pdf)
+
+## Post GSoC
+
+However, it's not enough. I have done some deep profiling about Clang, and found out the bottleneck:
+
+```
+% total        Total        Calls  Function
+   99.3  444'432'413       74'169  main [4]
+   92.3  412'998'941       29'027  llvm::SmallVectorBase::grow_pod(void*, unsigned long, unsigned long) [5]
+   89.2  399'211'718        7'870  clang::CompilerInstance::loadModule(clang::SourceLocation, llvm::ArrayRef<std::pair<clang::IdentifierInfo*, clang::SourceLocation> >, clang::Module::NameVisibilityKind, bool) [6]
+   89.2  399'058'838        7'644  clang::ASTReader::ReadAST(llvm::StringRef, clang::serialization::ModuleKind, clang::SourceLocation, unsigned int, llvm::SmallVectorImpl<clang::ASTReader::ImportedSubmodule>*) [7]
+   88.8  397'250'710        8'023  cling::Interpreter::loadModule(clang::Module*, bool) [8]
+   88.7  397'152'678        7'901  clang::Sema::ActOnModuleImport(clang::SourceLocation, clang::SourceLocation, clang::SourceLocation, llvm::ArrayRef<std::pair<clang::IdentifierInfo*, clang::SourceLocation> >) [9]
+   84.0  376'053'039       12'354  TRint::TRint(char const*, int*, char**, void*, int, bool, bool) [10]
+
+```
+
+From the above, we can see the hottest method is `SmallVectorBase::grow_pod`. That is actually caused by a problematic model: Clang always preallocates a huge buffer that is enough to contain all `SourceLocation` in a module, even without use.
+
+To further improve modules in ROOT, I plan to continue the work and cooperate with the upstream, and hopefully, we can solve this longstanding issue and make Clang better.
 
+## Other work
+
+Except the modules work, I also done an enormous contribution to `clang-repl`, which is an upstream version of ROOT's core interpreter. I improved the correctness and reliability of `clang-repl` by fixing lots of bugs. I also added an initial implementation of the code undo feature. Here're some noticeable patches:
+
+* https://reviews.llvm.org/D126781
+* https://reviews.llvm.org/D126682
+* https://reviews.llvm.org/D128782
+* https://reviews.llvm.org/D130420
+* https://reviews.llvm.org/D130831
+
+## It's not just about code!
+
+During this summer, I have learned a lot about coding stuff, and enhanced my knowledge of C++ and compiler. However, I think the most crucial thing I have learned has nothing to do with programming itself. In the last two months, I realized the importance of communication. That said, you can't just always code and debug, you need to understand your real goal, and that's achieved by communicating with your teammates. I also realized that my weakness -- not being meticulous enough. That really comes to bite me and makes me make some really stupid mistakes.
```

Comment:
> During this summer, I have learned a lot about coding stuff, and enhanced my knowledge of C++ and compiler. However, I think the most crucial thing I have learned has nothing to do with programming itself. In the last two months, I realized the importance of communication. That said, you can't just always code and debug, you need to understand your real goal, and that's achieved by communicating with your teammates.

That happens often when the codebases are huge and the knowledge is dispersed between various people/teams. Then that becomes more of finding the right communication channels to extract the knowledge and turn it into an acceptable solution. So fair amount of social engineering. I would turn this paragraph a bit around saying something around that rather than a self criticism.

> I also realized that my weakness -- not being meticulous enough. That really comes to bite me and makes me make some really stupid mistakes.

I think we should probably rework or drop these.